### PR TITLE
*: riscv fixes/changes

### PR DESCRIPTION
**Pull request by Madouura**
###### Description of changes
Add support for the RISC-V StarFive VisionFive SoC.
Sections of this taken from https://github.com/zhaofengli/nixos-riscv64.

Current goal(s)
- Fix known upstream bugs in other issues/PRs

Past goal(s)
- Be able to build a basic NixOS system for riscv64-linux and have it run on VisionFive hardware

Past related issue(s)
- #168813
  - Fixed by #168303
- #168829
  - Fixed by #166992
- #169278
  - Fixed by #170991

###### Things done

- Built on platform(s)
  - [x] x86_64-linux cross to riscv64-linux
  - [ ] aarch64-linux
  - [ ] riscv64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Madouura** [comment]
Logrotate seems to fail right at ``post-installation fixup`` for some... reason..
```bash
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/yl7lljq9mcqci0k7wlbns4xm8fjz6321-logrotate-riscv64-unknown-linux-gnu-3.19.0
shrinking /nix/store/yl7lljq9mcqci0k7wlbns4xm8fjz6321-logrotate-riscv64-unknown-linux-gnu-3.19.0/sbin/logrotate
gzipping man pages under /nix/store/yl7lljq9mcqci0k7wlbns4xm8fjz6321-logrotate-riscv64-unknown-linux-gnu-3.19.0/share/man/
riscv64-unknown-linux-gnu-strip is /nix/store/pmbl1qd8iqk4b44sdrj3advh8s9gcfbq-riscv64-unknown-linux-gnu-stage-final-gcc-debug-wrapper-10.3.0/bin/riscv64-unknown-linux-gnu-strip
stripping (with command riscv64-unknown-linux-gnu-strip and flags -S) in /nix/store/yl7lljq9mcqci0k7wlbns4xm8fjz6321-logrotate-riscv64-unknown-linux-gnu-3.19.0/sbin
patching script interpreter paths in /nix/store/yl7lljq9mcqci0k7wlbns4xm8fjz6321-logrotate-riscv64-unknown-linux-gnu-3.19.0
checking for references to /build/ in /nix/store/yl7lljq9mcqci0k7wlbns4xm8fjz6321-logrotate-riscv64-unknown-linux-gnu-3.19.0...
moving /nix/store/yl7lljq9mcqci0k7wlbns4xm8fjz6321-logrotate-riscv64-unknown-linux-gnu-3.19.0/sbin/* to /nix/store/yl7lljq9mcqci0k7wlbns4xm8fjz6321-logrotate-riscv64-unknown-linux-gnu-3.19.0/bin
building '/nix/store/n3nq3d79cvsf98b5pw1z88jn74xp43fv-logrotate.conf.drv'...
/build/.attr-0: line 14: /nix/store/76ry6fv0c7n31ah9jg0jm3mqci2n5knb-coreutils-riscv64-unknown-linux-gnu-9.0/bin/id: cannot execute binary file: Exec format error
builder for '/nix/store/n3nq3d79cvsf98b5pw1z88jn74xp43fv-logrotate.conf.drv' failed with exit code 126
cannot build derivation '/nix/store/a02x84vy2bak2b78m5vasdik8sd3nbf8-unit-logrotate-checkconf.service.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/cjzi5xqb95swyk490c3fv5jy0p4fwzw5-unit-logrotate.service.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/vacc5bn1szk99p3pyb2q220xphyl6srj-system-units.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/wh3v282827cxpa4zfmlp5j7mrgxrjhk0-etc.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/x24y87d0lpdz6ql2yfv3gfhkyfmabr6x-nixos-system-nixos-22.05pre-git.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/wp1h969krwbwlfdcqnrf106cgx062x8m-ext4-fs.img.zst-riscv64-unknown-linux-gnu.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/7rv1nz4ksch76g316s6wqpcghqzxklxl-nixos-sd-image-22.05pre-git-riscv64-linux-visionfive.img-riscv64-unknown-linux-gnu.drv': 1 dependencies couldn't be built
error: build of '/nix/store/7rv1nz4ksch76g316s6wqpcghqzxklxl-nixos-sd-image-22.05pre-git-riscv64-linux-visionfive.img-riscv64-unknown-linux-gnu.drv' failed
```

**Comment by Madouura** [comment]
The sd-image changes may be able to be genericized if multiple other systems also use a similar sdcard layout.

**Comment by Madouura** [comment]
Would have liked to use the starfive linux fork as a patch on top of our kernel instead, but it would always fail at the patch stage, so instead using it verbatim for now.

**Comment by colemickens** [comment]
cc: @zhaofengli

**Comment by Madouura** [comment]
...Now that I think about it, should probably be targeting staging as this is a HUGE rebuild for hydra lol.

**Comment by Madouura** [comment]
Should also mention I'm holding off on release notes until everything is properly building and I can test it on my hardware.
Currently, with an iso.nix (such as below) and ``nix-build nixos -A config.system.build.sdImage -I nixos-config=iso.nix`` it does build and generate an iso.
```nix
{ config, lib, pkgs, ... }:

{
  imports = [ "/home/mado/Documents/Development/nixpkgs/nixos/modules/installer/sd-card/sd-image-riscv64-visionfive.nix" ];

  nixpkgs = {
    config.allowUnsupportedSystem = true;

    crossSystem = {
      config = "riscv64-unknown-linux-gnu";
      system = "riscv64-linux";
    };
  };
}
```

**Comment by Madouura** [comment]
Geez... chill out git, I just wanted to rebase.

File: pkgs/os-specific/linux/kernel/linux-visionfive.nix
Comment (zhaofengli) [inline pkgs/os-specific/linux/kernel/linux-visionfive.nix:36]: Could you add @ius (`ius`) and me (`zhaofengli`) as maintainers? Would also appreciate co-authorship in the commits as well.

File: nixos/modules/installer/sd-card/sd-image.nix
Comment (zhaofengli) [inline nixos/modules/installer/sd-card/sd-image.nix:212]: I'm a little bit wary of adding this firmware-specific hack to mainline nixpkgs, since I expect this to be fixed in the future releases of the vendor U-Boot. At the very least some explanation should be added (https://github.com/starfive-tech/u-boot/issues/30).

Comment (Madouura) [reply nixos/modules/installer/sd-card/sd-image.nix:212]: That's one of the reasons I made this change it's own commit as well, so we could easily revert it.
It is needed, but I'll add some explanation like you ask.

File: pkgs/os-specific/linux/kernel/linux-visionfive.nix
Comment (Madouura) [reply pkgs/os-specific/linux/kernel/linux-visionfive.nix:36]: Sure, I don't mind.

**Comment by zhaofengli** [comment]
Also cc'ing @samueldr for the sd-image stuff.

**Comment by Madouura** [comment]
Temporarily reverting #121147 in this PR to fix man-db build.

**Comment by Madouura** [comment]
Should be able to test tomorrow once I get that HDMI cable.

**Comment by Mindavi** [comment]
Man-db cross should be fixed here: https://github.com/NixOS/nixpkgs/pull/169380

**Comment by samueldr** [comment]
My opinion is that we should not add device-specific image generation, and strive for standards-based booting. E.g. UEFI, though extlinux.conf-compatible is *adequate enough* for the time being.

Though this is a bit harder to resolve given the issue that the device requires a device-specific kernel. Again my opinion is that Nixpkgs shouldn't have vendor-specific kernels.

These are my opinions. Let's see what others think about that.

I'm not sure what is the best way forward to handle the device-specific kernels problem.

* * *

Note that for standards-based boot, I'll be glad to work with y'all to get a Tow-Boot build together (see https://github.com/Tow-Boot/Tow-Boot/pull/145)

**Comment by Madouura** [comment]
temporarily added man-db cross compilation fixes.

**Comment by Madouura** [comment]
Using
```nix
{ config, ... }:

{
  imports = [ "/home/mado/Documents/Development/nixpkgs/nixos/modules/installer/sd-card/sd-image-riscv64-visionfive-installer.nix" ];

  nixpkgs.crossSystem = {
    config = "riscv64-linux";
    system = "riscv64-linux";
  };
}
```

**Comment by Madouura** [comment]
Anyway, it does work on the hardware, and I was able to do some stuff via hdmi.
Since a rebase to the latest master, I get an invalid input problem for my screen, but I'm hoping that'll fix itself whenever I rebase again.
Don't have a lot of time, so I'll have to leave further testing with ssh for Monday.

**Comment by Madouura** [comment]
Also note for future:
We're going to have a hell of a time on our hands getting desktop managers and display managers to work with this, for one, boost isn't supported, so that rules out gdm and gnome for the meantime. Lightdm has some compilation errors, etc.

**Comment by Madouura** [comment]
> My opinion is that we should not add device-specific image generation, and strive for standards-based booting. E.g. UEFI, though extlinux.conf-compatible is _adequate enough_ for the time being.
> 
> Though this is a bit harder to resolve given the issue that the device requires a device-specific kernel. Again my opinion is that Nixpkgs shouldn't have vendor-specific kernels.
> 
> These are my opinions. Let's see what others think about that.
> 
> I'm not sure what is the best way forward to handle the device-specific kernels problem.
> 
> Note that for standards-based boot, I'll be glad to work with y'all to get a Tow-Boot build together (see [Tow-Boot/Tow-Boot#145](https://github.com/Tow-Boot/Tow-Boot/pull/145))

I completely agree, I don't personally think this PR should be merged until we either get u-boot fixed or tow-boot going.

**Comment by NickCao** [comment]
I managed to get u-boot fixed https://github.com/NickCao/u-boot-starfive/commit/d5f53001d7324c9680ab39c8a4c87ac03a4162b8, thanks @zhaofengli for pointing the direction.

**Comment by Madouura** [comment]
> I managed to get u-boot fixed [NickCao/u-boot-starfive@d5f5300](https://github.com/NickCao/u-boot-starfive/commit/d5f53001d7324c9680ab39c8a4c87ac03a4162b8), thanks @zhaofengli for pointing the direction.

Oh my. I'll have to test this ASAP!

**Comment by Madouura** [comment]
Having some issues with cross-compiling u-boot... Not entirely sure what to do, same issue as with linux in the current master, but even changed stdenv to use gcc11 isn't working.

**Comment by NickCao** [comment]
u-boot and opensbi is broken by binutils 2.38, you can find the additional patches requires at https://github.com/NickCao/nixos-riscv. Additionally, there is prebuilt binary at https://hydra.nichi.co/build/1321930 that can be directly flashed.

**Comment by Madouura** [comment]
> u-boot and opensbi is broken by binutils 2.38, you can find the additional patches requires at https://github.com/NickCao/nixos-riscv. Additionally, there is prebuilt binary at https://hydra.nichi.co/build/1321930 that can be directly flashed.

Oh so it was binutils after all. There's a temporary workaround using ``NIX_CFLAGS_COMPILE = "-march=rv64gc_zicsr_zifencei";`` in u-boot's default.nix in the meantime.

File: pkgs/misc/uboot/default.nix
Comment (NickCao) [inline pkgs/misc/uboot/default.nix:511]: This should be starfive_jh7100_visionfive_smode_defconfig

Comment (Madouura) [reply pkgs/misc/uboot/default.nix:511]: I was looking for that in the u-boot repo but never found it, it compiles but are you sure it's valid?

Comment (NickCao) [reply pkgs/misc/uboot/default.nix:511]: It's here https://github.com/NickCao/u-boot-starfive/blob/master/configs/starfive_jh7100_visionfive_smode_defconfig

Comment (NickCao) [reply pkgs/misc/uboot/default.nix:511]: My u-boot tree is from the starfive fork base onto u-boot v2022.4, upstream u-boot has no support for visionfive yet.

Comment (Madouura) [reply pkgs/misc/uboot/default.nix:511]: Ahh okay, thanks.

**Comment by Madouura** [comment]
Hm, the board doesn't seem to like flashing u-boot, even with older official versions.

**Comment by NickCao** [comment]
> Hm, the board doesn't seem to like flashing u-boot, even with older official versions.

try `picocom --send-cmd "sx -vv"`

**Comment by Madouura** [comment]
Had to use ``/nix/store/sjl2ykvxr3cly297mhg3d538z3c8vpl4-lrzsz-0.12.20/bin/sx`` specifically.
Maybe I was doing it wrong by using ``nix-shell``, but picocom and minicom isn't particularly intuitive when it comes to this.

**Comment by Madouura** [comment]
Anyway, just werks now, thank you Nick!

**Comment by NickCao** [comment]
There's much duplicated code in `nixos/modules/tasks/visionfive.nix` and `nixos/modules/installer/sd-card/sd-image-riscv64-visionfive.nix`, and I'm not seeing hardware specific configurations for other hardware in `nixos/module/tasks`,  shall we remove the former altogether.

**Comment by Madouura** [comment]
> There's much duplicated code in `nixos/modules/tasks/visionfive.nix` and `nixos/modules/installer/sd-card/sd-image-riscv64-visionfive.nix`, and I'm not seeing hardware specific configurations for other hardware in `nixos/module/tasks`, shall we remove the former altogether.
I'm not entirely sure it even fires, so sure.

File: pkgs/misc/uboot/default.nix
Comment (Madouura) [inline pkgs/misc/uboot/default.nix:72]: I forgot about that, I'll remove it next update.

Comment (Madouura) [reply pkgs/misc/uboot/default.nix:72]: actually, I'll just do it now...

File: pkgs/misc/bootrom-visionfive/default.nix
Comment (NickCao) [inline pkgs/misc/bootrom-visionfive/default.nix:6]: `bootrom` is not the perfect name for this complex as it actually means the on chip readonly first stage bootloader. And in starfive docs this is called `fw_payload.bin.out` or erroneously, `u-boot`. Simply `firmware` can be a better choice.

File: nixos/modules/installer/sd-card/sd-image-riscv64-visionfive.nix
Comment (NickCao) [inline nixos/modules/installer/sd-card/sd-image-riscv64-visionfive.nix:15]: Already fixed in https://github.com/NixOS/nixpkgs/pull/166992

File: nixos/modules/installer/sd-card/sd-image-riscv64-visionfive.nix
Comment (NickCao) [inline nixos/modules/installer/sd-card/sd-image-riscv64-visionfive.nix:44]: What's the purpose of this?

File: pkgs/misc/bootrom-visionfive/default.nix
Comment (NickCao) [inline pkgs/misc/bootrom-visionfive/default.nix:8]: No need to add it as a buildInput.

File: nixos/modules/installer/sd-card/sd-image-riscv64-visionfive.nix
Comment (Madouura) [reply nixos/modules/installer/sd-card/sd-image-riscv64-visionfive.nix:44]: To be able to use this with the hardware, we have the option of flashing from microSD, thus this seems a fitting addition.

File: pkgs/misc/bootrom-visionfive/default.nix
Comment (Madouura) [reply pkgs/misc/bootrom-visionfive/default.nix:6]: Good point.

File: nixos/modules/installer/sd-card/sd-image-riscv64-visionfive.nix
Comment (Madouura) [reply nixos/modules/installer/sd-card/sd-image-riscv64-visionfive.nix]: Specifically, we can use ``load`` to load a binary from the sd card, and use ``sf update`` to overwrite the flash.

**Comment by NickCao** [comment]
Just found out that  gcc11.3 and binutils 2.38 is a match in heaven and the u-boot patch is no longer needed.

**Comment by Madouura** [comment]
Niiice.
I'll remove the patch once #170121 and #170991 is in master.

**Comment by Madouura** [comment]
Just got home and I've been so far, successfully compiling the entire GCC build chain and who knows what else with nixos-rebuild natively, so there's a small note of progress.

**Comment by Madouura** [comment]
``man-db`` and ``groff`` commits will no longer be needed once #169965 is merged into master.

**Comment by Madouura** [comment]
If I'm not mistaken, GRUB2 support for RISC-V should be in-place.
Are we waiting for a GRUB2 update to allow for enabling GRUB for boards like this?

**Comment by NickCao** [comment]
> If I'm not mistaken, GRUB2 support for RISC-V should be in-place. Are we waiting for a GRUB2 release to allow for enabling GRUB for boards like this?

But we still have to chainload grub from u-boot, right? That makes little sense to me.

**Comment by Madouura** [comment]
Yes, it was a more general question, slightly off-topic from this PR.
I just noticed we have GRUB disabled for most of the non-x86 sdcard images, so thus the question.

**Comment by Madouura** [comment]
Actually, GRUB is disabled for all of them.
https://search.nixos.org/options?channel=21.11&show=boot.loader.grub.enable&from=0&size=50&sort=relevance&type=packages&query=grub
Seems like GRUB is enabled by default.

**Comment by Madouura** [comment]
Yeah, this is off-topic.
Let's move this to #171133

**Comment by Madouura** [comment]
coreutils fails in latest rebase at the test stage, seems to be ``tests/misc/shred-exact.sh``.

**Comment by Madouura** [comment]
libuv also fails the test phase, for whatever reason.

**Comment by Madouura** [comment]
Too many problems related to changes made for 22.05, gonna wait maybe a week after that's merged and then continue working on this.

**Comment by Madouura** [comment]
Builds cleanly once again after the latest staging-next merge.
Now just gotta wait till I get home to start building on the hardware itself again.

**Comment by Madouura** [comment]
Hm, there are issues compiling zstd, I'll have to investigate.

**Comment by NickCao** [comment]
> Hm, there are issues compiling (linking?) zstd, I'll have to investigate.

See https://github.com/NickCao/nixpkgs/commit/90ae9187f3b2dc154b011d15ad6d31678bced606

**Comment by Madouura** [comment]
> > Hm, there are issues compiling (linking?) zstd, I'll have to investigate.
> 
> See [NickCao@90ae918](https://github.com/NickCao/nixpkgs/commit/90ae9187f3b2dc154b011d15ad6d31678bced606)

Ah, that's already been figured out.
Would you mind submitting a PR regarding this? I'll add that fix as a temporary fix to the commits.

**Comment by NickCao** [comment]
> Would you mind submitting a PR regarding this? I'll add that fix as a temporary fix to the commits.

We should fix this upstream instead of simply disable failing parts, let's keep this a *temporary* fix.

**Comment by Madouura** [comment]
Good point.

**Comment by Madouura** [comment]
@NickCao Is there any reason to not include https://github.com/NickCao/nixpkgs/commit/871bd07e4276342155a5050cce83edb4ecd16b77 in this PR?
The goal is a native ``nixos-rebuild`` and being able to generate it's own image, so I _think_ it makes sense.

**Comment by Madouura** [comment]
Alright, I'm fairly confident the only remaining problems are things that need to be fixed upstream.
libffi: tests need to be fixed for riscv upstream.
zstd: contrib building needs to be fixed for riscv upstream. (pzstd linker error) 
nlohmann_json: tests need to be fixed for riscv upstream.
aws-sdk-cpp: tests need to be fixed for riscv upstream. (linker error)

**Comment by Madouura** [comment]
I'll open new issues for the aforementioned upstream problems as I compile and get the proper error logs on hardware.
In the meantime, I believe this is ready for review.

**Comment by Madouura** [comment]
Looks like ``nlohmann_json`` and ``aws-sdk-cpp`` issues will have to be put on hold until ``libffi`` and ``zstd`` are fixed.

File: nixos/modules/installer/sd-card/sd-image-riscv64-visionfive.nix
Comment (06kellyjac) [inline nixos/modules/installer/sd-card/sd-image-riscv64-visionfive.nix:45]: how does one flash this custom firmware? Do I need to follow this [VisionFive Quick Start Guide - Appendix B Updating Firmware and U Boot](https://rvspace.org/en/Product/VisionFive/Technical_Documents/VisionFive_Single_Board_Computer_Quick_Start_Guide#appendix-b-updating-firmware-and-u-boot) or is there an easier way?

Comment (06kellyjac) [inline nixos/modules/installer/sd-card/sd-image-riscv64-visionfive.nix:45]: how does one flash this custom firmware? Do I need to follow this [VisionFive Quick Start Guide - Appendix B Updating Firmware and U Boot](https://rvspace.org/en/Product/VisionFive/Technical_Documents/VisionFive_Single_Board_Computer_Quick_Start_Guide#appendix-b-updating-firmware-and-u-boot) or is there an easier way?

Comment (Madouura) [reply nixos/modules/installer/sd-card/sd-image-riscv64-visionfive.nix:45]: Yep, that's it.
I found it easier to update it via serial with xmodem with picocom.

**Comment by MatthewCroughan** [comment]
The image generated here doesn't boot on the VisionFive hardware, whereas https://github.com/zhaofengli/nixos-riscv64/releases does.

**Comment by MatthewCroughan** [comment]
@NickCao the image from Zhaofeng does not require the same procedure to work. So what is the reason that we require a u-boot flashing procedure specifically for this PR?

**Comment by NickCao** [comment]
The images from Fedora and Zhaofeng both follows a very specific partition layout as enforced by the factory u-boot. While our patched u-boot works with all generic images following the distro boot specification.

**Comment by MatthewCroughan** [comment]
@NickCao is this the specification you're referring to? https://developer.toradex.com/linux-bsp/how-to/boot/distro-boot/

**Comment by NickCao** [comment]
> @NickCao is this the specification you're referring to? https://developer.toradex.com/linux-bsp/how-to/boot/distro-boot/

yes, actually you can just consider it as `ext-linux-compatible`.

**Comment by samueldr** [comment]
The correct documentation is the one from upstream U-Boot:

 - https://source.denx.de/u-boot/u-boot/-/blob/v2022.04/doc/develop/distro.rst

**Comment by zhaofengli** [comment]
The patched U-Boot is the way to go until upstream adopts https://github.com/starfive-tech/u-boot/pull/31. See also:

- https://github.com/starfive-tech/u-boot/issues/30
- https://github.com/NixOS/nixpkgs/pull/168826#discussion_r852381630

**Comment by luizribeiro** [comment]
Just wanted to report that I was able to successfully boot NixOS on my VisionFive board today with this PR. Everything I tested works fine. I didn't attempt using video, as I have been accessing my board through the serial port. Would love to see this merged!

**Comment by Madouura** [comment]
I personally have had issues trying to get anything to display with the HDMI port, usually the board not transmitting the correct EDID for whatever reason, so my monitor won't display anything.
I don't know where to begin with that, but if you find anything out, let me know.

**Comment by luizribeiro** [comment]
@Madouura, I had similar problems with HDMI on Fedora - which is why I haven't been using it. I'll give it a try later.

FWIW I was able to get Wifi working by applying an overlay which copies `brcmfmac43430-sdio.AP6212.txt` as `brcmfmac43430-sdio.starfive,visionfive-v1.txt`:

```
nixpkgs.overlays = [
  (pkgs: prev: {
    linux-firmware = prev.linux-firmware.overrideAttrs (old: {
      postInstall = ''
        cp $out/lib/firmware/brcm/brcmfmac43430-sdio.AP6212.txt \
          $out/lib/firmware/brcm/brcmfmac43430-sdio.starfive,visionfive-v1.txt
      '';
      outputHash = null;
    });
  })
];
```

There is an [upcoming upstream patch](https://lore.kernel.org/all/CADWks+YJm8bi+KPXYTvQ3JrriDW2dcdxfSZ2O5J0vfhfC654Tw@mail.gmail.com/) to `linux-firmware` for this though - so not sure if it is worth it getting a package for this firmware into `nixpkgs` (or patching `linux-firmware` ourselves).

**Comment by Madouura** [comment]
>There is an [upcoming upstream patch](https://lore.kernel.org/all/CADWks+YJm8bi+KPXYTvQ3JrriDW2dcdxfSZ2O5J0vfhfC654Tw@mail.gmail.com/) to linux-firmware for this though - so not sure if it is worth it getting a package for this firmware into nixpkgs (or patching linux-firmware ourselves).

Considering how niche this is already, we probably shouldn't include any specializations such as that yet. If this were more widely used then it would be justified, but in this case, let's wait for upstream.

**Comment by alyssais** [comment]
I agree with @samueldr that we shouldn't be adding new vendor kernels to Nixpkgs.  We have [nixos-hardware](https://github.com/NixOS/nixos-hardware/) for that sort of thing.  The firmware package is probably okay for Nixpkgs though, and opensbi.

**Comment by samueldr** [comment]
Yes, having *stock unmodified* upstream U-Boot [and tooling] is acceptable and desirable in Nixpkgs. As long as it's understood that it's not meant to be automatically managed by the lifecycle of NIxOS (e.g. no auto-upgrades).

**Comment by Madouura** [comment]
> I agree with @samueldr that we shouldn't be adding new vendor kernels to Nixpkgs. We have [nixos-hardware](https://github.com/NixOS/nixos-hardware/) for that sort of thing. The firmware package is probably okay for Nixpkgs though, and opensbi.

OpenSBI-master is likely not needed anymore due to a new version release, I have no tested it yet though.

**Comment by Madouura** [comment]
A couple steps away from a successful ``nixos-rebuild`` on RISC-V.
https://gist.github.com/Madouura/edf2ac1b06bf6e63507b9dc9bb338064
Seems to be an issue with the unicode test.

**Comment by Madouura** [comment]
Looks like LLVM 14 is failing it's unit tests.

**Comment by luizribeiro** [comment]
@Madouura, I was able to get past those LLVM test failures with this overlay: https://github.com/luizribeiro/nixpkgs-riscv/blob/main/overlays/host.nix#L152-L180

I'll put up a PR for this.

**Comment by luizribeiro** [comment]
#181798 should fix those tests.

**Comment by Madouura** [comment]
It works beautifully. LLVM-14 has successfully built on hardware.
Just Rust and a few other packages and we're golden.

**Comment by MatthewCroughan** [comment]
I've made the following repository and youtube video which is largely based on the work from this PR. It describes how to flash our bootloader to the VisionFive board, and provides a script for doing it `nix run github:matthewcroughan/visionfive-nix#flashBootloader`

https://youtu.be/cmjK8SLIx9E
https://github.com/MatthewCroughan/visionfive-nix

**Comment by Madouura** [comment]
Thank you!

**Comment by Madouura** [comment]
Keeping this all-in-one until I can get to implementing in nixos-hardware.

**Comment by Madouura** [comment]
Moved mostly everything to https://github.com/NixOS/nixos-hardware/pull/479

**Comment by Madouura** [comment]
Are we good for merge?

**Comment by SuperSandro2000** [comment]
@astro FYI

File: pkgs/development/libraries/aws-sdk-cpp/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/aws-sdk-cpp/default.nix:90]: This is missing a short comment.

File: pkgs/development/libraries/nlohmann_json/default.nix
Comment (SuperSandro2000) [inline pkgs/development/libraries/nlohmann_json/default.nix:28]: presumable this skips some slow tests? Are they valuable for us? I assume if we skip them for all platforms they are not super important.

Comment (Madouura) [reply pkgs/development/libraries/nlohmann_json/default.nix:28]: For ``JSON_FASTTESTS``: https://github.com/nlohmann/json/blob/dae54122684bbe256117b82992e95e7840720b75/tests/CMakeLists.txt#L4
Just skips some expensive and slow tests. I have no idea if they're valuable to us or not, but I doubt it.
For ``JSON_BUILDTESTS``: IIRC this is just the actually correct flag to build tests.

Comment (06kellyjac) [reply pkgs/development/libraries/nlohmann_json/default.nix:28]: It'd be interesting to know what the speed increases are here

Comment (Madouura) [reply pkgs/development/libraries/nlohmann_json/default.nix:28]: I do know it's required for risc-v on visionfive at least, I can just make it a ``lib.optionals`` if that's preferable.

Comment (SuperSandro2000) [reply pkgs/development/libraries/nlohmann_json/default.nix:28]: > It'd be interesting to know what the speed increases are here

Fast tests take ~2 seconds, slow tests are ~120-190 sec plus building them. I would vote to always only run the fast tests.

**Comment by Madouura** [comment]
Assuming the ``nlohmann_json`` change is okay, we should be good for merge.

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by SuperSandro2000** [comment]
@ofborg build nix

Merging when nix still builds on the major platforms.

**Pull request merged**
